### ISSUE TITLE
mixin: Grafana Dashboards fail server side apply

### ISSUE THREAD

**Issue by itspngu** [thanos-io/thanos#4880]
**Thanos, Prometheus and Golang version used**:

quay.io/thanos/thanos:v0.22.0

**Object Storage Provider**:

GCS

**What happened**:

When using the Thanos mixin with kube-prometheus, the resulting yaml manifests fail server side apply (and its validation). In particular, `grafana-dashboard-bucket_replicate` and `grafana-dashboard-query_frontend` are violating a naming convention, the rest is fine. This makes using the mixin as part of a GitOps workflow impossible, see also https://github.com/fluxcd/kustomize-controller/issues/491#issuecomment-973988095:

```
Error from server (Invalid): ConfigMap "grafana-dashboard-bucket_replicate" is invalid: metadata.name: Invalid value: "grafana-dashboard-bucket_replicate": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
Error from server (Invalid): ConfigMap "grafana-dashboard-query_frontend" is invalid: metadata.name: Invalid value: "grafana-dashboard-query_frontend": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

**What you expected to happen**:

Generation of valid yaml manifests that can be applied.

**How to reproduce it (as minimally and precisely as possible)**:

I tried to set up a minimal test case for half an hour and got fed up with it, sorry. Just try to use the mixin with kube-prometheus and server side apply. See the logs below and the linked issue for additional context.

**Full logs to relevant components**:

<details>Logs
<p>

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.6", GitCommit:"d921bc6d1810da51177fbd0ed61dc811c5228097", GitTreeState:"clean", BuildDate:"2021-10-27T17:50:34Z", GoVersion:"go1.16.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.5-gke.1802", GitCommit:"d464a255a4acb8dca1e0914586f4f5085681fb51", GitTreeState:"clean", BuildDate:"2021-10-21T21:43:23Z", GoVersion:"go1.16.7b7", Compiler:"gc", Platform:"linux/amd64"}

$ kubeval kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains an empty YAML document
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-alertmanager-overview)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-apiserver)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-bucket_replicate)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-cluster-total)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-compact)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-controller-manager)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-k8s-resources-cluster)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-k8s-resources-namespace)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-k8s-resources-node)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-k8s-resources-pod)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-k8s-resources-workload)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-k8s-resources-workloads-namespace)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-kubelet)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-namespace-by-pod)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-namespace-by-workload)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-node-cluster-rsrc-use)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-node-rsrc-use)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-nodes)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-overview)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-persistentvolumesusage)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-pod-total)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-prometheus-remote-write)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-prometheus)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-proxy)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-query)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-query_frontend)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-receive)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-rule)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-scheduler)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-sidecar)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-store)
PASS - kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml contains a valid ConfigMap (monitoring.grafana-dashboard-workload-total)

$ kubectl -n monitoring apply -f kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml --dry-run=client
configmap/grafana-dashboard-alertmanager-overview created (dry run)
configmap/grafana-dashboard-apiserver created (dry run)
configmap/grafana-dashboard-bucket_replicate created (dry run)
configmap/grafana-dashboard-cluster-total created (dry run)
configmap/grafana-dashboard-compact created (dry run)
configmap/grafana-dashboard-controller-manager created (dry run)
configmap/grafana-dashboard-k8s-resources-cluster created (dry run)
configmap/grafana-dashboard-k8s-resources-namespace created (dry run)
configmap/grafana-dashboard-k8s-resources-node created (dry run)
configmap/grafana-dashboard-k8s-resources-pod created (dry run)
configmap/grafana-dashboard-k8s-resources-workload created (dry run)
configmap/grafana-dashboard-k8s-resources-workloads-namespace created (dry run)
configmap/grafana-dashboard-kubelet created (dry run)
configmap/grafana-dashboard-namespace-by-pod created (dry run)
configmap/grafana-dashboard-namespace-by-workload created (dry run)
configmap/grafana-dashboard-node-cluster-rsrc-use created (dry run)
configmap/grafana-dashboard-node-rsrc-use created (dry run)
configmap/grafana-dashboard-nodes created (dry run)
configmap/grafana-dashboard-overview created (dry run)
configmap/grafana-dashboard-persistentvolumesusage created (dry run)
configmap/grafana-dashboard-pod-total created (dry run)
configmap/grafana-dashboard-prometheus-remote-write created (dry run)
configmap/grafana-dashboard-prometheus created (dry run)
configmap/grafana-dashboard-proxy created (dry run)
configmap/grafana-dashboard-query created (dry run)
configmap/grafana-dashboard-query_frontend created (dry run)
configmap/grafana-dashboard-receive created (dry run)
configmap/grafana-dashboard-rule created (dry run)
configmap/grafana-dashboard-scheduler created (dry run)
configmap/grafana-dashboard-sidecar created (dry run)
configmap/grafana-dashboard-store created (dry run)
configmap/grafana-dashboard-workload-total created (dry run)

kubectl -n monitoring apply -f kubernetes/clusters/management/kube-prometheus/components/grafana-dashboardDefinitions.yaml --dry-run=server
configmap/grafana-dashboard-alertmanager-overview created (server dry run)
configmap/grafana-dashboard-apiserver created (server dry run)
configmap/grafana-dashboard-cluster-total created (server dry run)
configmap/grafana-dashboard-compact created (server dry run)
configmap/grafana-dashboard-controller-manager created (server dry run)
configmap/grafana-dashboard-k8s-resources-cluster created (server dry run)
configmap/grafana-dashboard-k8s-resources-namespace created (server dry run)
configmap/grafana-dashboard-k8s-resources-node created (server dry run)
configmap/grafana-dashboard-k8s-resources-pod created (server dry run)
configmap/grafana-dashboard-k8s-resources-workload created (server dry run)
configmap/grafana-dashboard-k8s-resources-workloads-namespace created (server dry run)
configmap/grafana-dashboard-kubelet created (server dry run)
configmap/grafana-dashboard-namespace-by-pod created (server dry run)
configmap/grafana-dashboard-namespace-by-workload created (server dry run)
configmap/grafana-dashboard-node-cluster-rsrc-use created (server dry run)
configmap/grafana-dashboard-node-rsrc-use created (server dry run)
configmap/grafana-dashboard-nodes created (server dry run)
configmap/grafana-dashboard-overview created (server dry run)
configmap/grafana-dashboard-persistentvolumesusage created (server dry run)
configmap/grafana-dashboard-pod-total created (server dry run)
configmap/grafana-dashboard-prometheus-remote-write created (server dry run)
configmap/grafana-dashboard-prometheus created (server dry run)
configmap/grafana-dashboard-proxy created (server dry run)
configmap/grafana-dashboard-query created (server dry run)
configmap/grafana-dashboard-receive created (server dry run)
configmap/grafana-dashboard-rule created (server dry run)
configmap/grafana-dashboard-scheduler created (server dry run)
configmap/grafana-dashboard-sidecar created (server dry run)
configmap/grafana-dashboard-store created (server dry run)
configmap/grafana-dashboard-workload-total created (server dry run)
Error from server (Invalid): ConfigMap "grafana-dashboard-bucket_replicate" is invalid: metadata.name: Invalid value: "grafana-dashboard-bucket_replicate": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
Error from server (Invalid): ConfigMap "grafana-dashboard-query_frontend" is invalid: metadata.name: Invalid value: "grafana-dashboard-query_frontend": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')

```

</p>
</details>

**Comment by itspngu** [comment]
It seems this has been fixed before... https://github.com/thanos-io/thanos/commit/2ed64cab2fde4ebb8674fe68eaaff381739011a0#diff-8b2bee54f896c2326a2a35a5f6ae1c404ef2d503c2186f48fa7bf810cf151e62

... and then broken again https://github.com/thanos-io/thanos/commit/569a4d730e5f41ca949d1f973671743da482de47#diff-61d624f57144d877c035b5fe65277dff301be7e52707e37fa538ad8750ed5a75

**Comment by itspngu** [comment]
I forgot to mark this as fixed in #4883. It's fixed. :)
